Move more layouting logic to `rustc_abi`

moulins opened the pull request:
Move all `LayoutData`-constructing code to `rustc_abi`:
- Infaillible operations get a new `LayoutData` constructor method;
- Faillible ones get a new method on `LayoutCalculator`.

moulins committed:
Remove most manual LayoutData creations and move them to `rustc_abi`

...either as:
- methods on LayoutCalculator, for faillible operations;
- constructors on LayoutData, for infaillible ones.

workingjubilee commented on compiler/rustc_abi/src/layout.rs:
```diff
@@ -102,41 +120,109 @@ impl<Cx: HasDataLayout> LayoutCalculator<Cx> {
-            .into_iter()
-            .chain(Niche::from_scalar(dl, Size::ZERO, a))
-            .max_by_key(|niche| niche.available(dl));
+    pub fn simd_type<
+        FieldIdx: Idx,
+        VariantIdx: Idx,
+        F: AsRef<LayoutData<FieldIdx, VariantIdx>> + fmt::Debug,
+    >(
+        &self,
+        element: F,
+        count: u64,
+        repr_packed: bool,
+    ) -> LayoutCalculatorResult<FieldIdx, VariantIdx, F> {
+        let elt = element.as_ref();
+        if count == 0 {
+            return Err(LayoutCalculatorError::ZeroLengthSimdType);
+        } else if count > MAX_SIMD_LANES {
+            return Err(LayoutCalculatorError::OversizedSimdType { max_lanes: MAX_SIMD_LANES });
+        }
+
+        let BackendRepr::Scalar(e_repr) = elt.backend_repr else {
+            return Err(LayoutCalculatorError::NonPrimitiveSimdType(element));
+        };
```
is this a direct copy of the logic in rust-analyzer? it's wrong.

workingjubilee replied:
answer to jubilee: it's not, the commit was just not split and you haven't finished your coffee yet.

moulins committed:
Move SIMD layout logic to `rustc_abi`

moulins committed:
Use `rustc_abi` code for SIMD layout in rust-analyzer

moulins committed:
Refactor coroutine layout logic to precompute all sublayouts

Also properly attaches spans on layouts of non-promoted coroutine
locals, which slightly improves the error messages for some coroutine tests.

moulins committed:
Move coroutine layout logic to `rustc_abi`

workingjubilee commented:
I walked through the commits checking to see if they passed `x.py check` and while that's not a strict requirement, there was a small change that a later commit added that fixed things, and it built after I moved that back, so I did.

workingjubilee commented on compiler/rustc_abi/src/layout.rs:
```diff
@@ -102,41 +120,109 @@ impl<Cx: HasDataLayout> LayoutCalculator<Cx> {
+        })
+    }
+
+    pub fn coroutine<
```
I'd like to see a doc comment on this which has two purposes, one is
- direct people to look at the associated module
- explain, *very briefly*, why this is its own function instead of coroutines being "erased" and transformed into "just an enum" or "just a struct" in layout terms by this point (one answer would be "because of their unique field-offset constraints")

workingjubilee commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -275,6 +275,13 @@ pub enum TargetDataLayoutErrors<'a> {
+    /// This value is selected based on backend support:
+    /// * LLVM does not appear to have a vector width limit.
+    /// * Cranelift stores the base-2 log of the lane count in a 4 bit integer.
+    pub const MAX_SIMD_LANES: u64 = 1 << 0xF;
```
Please do not make this into an associated constant. It is not actually dependent on the target.

workingjubilee commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -1745,6 +1752,42 @@ impl<FieldIdx: Idx, VariantIdx: Idx> LayoutData<FieldIdx, VariantIdx> {
         self.uninhabited
     }
 
+    pub fn unit<C: HasDataLayout>(cx: &C, sized: bool) -> Self {
+        let dl = cx.data_layout();
+        LayoutData {
+            variants: Variants::Single { index: VariantIdx::new(0) },
+            fields: FieldsShape::Arbitrary {
+                offsets: IndexVec::new(),
+                memory_index: IndexVec::new(),
+            },
+            backend_repr: BackendRepr::Memory { sized },
+            largest_niche: None,
+            uninhabited: false,
+            align: dl.i8_align,
+            size: Size::ZERO,
+            max_repr_align: None,
+            unadjusted_abi_align: dl.i8_align.abi,
+            randomization_seed: Hash64::new(0),
+        }
+    }
+
+    pub fn never_type<C: HasDataLayout>(cx: &C) -> Self {
+        let dl = cx.data_layout();
+        // This is also used for uninhabited enums, so we use `Variants::Empty`.
+        LayoutData {
+            variants: Variants::Empty,
+            fields: FieldsShape::Primitive,
+            backend_repr: BackendRepr::Memory { sized: true },
+            largest_niche: None,
+            uninhabited: true,
+            align: dl.i8_align,
+            size: Size::ZERO,
+            max_repr_align: None,
+            unadjusted_abi_align: dl.i8_align.abi,
+            randomization_seed: Hash64::ZERO,
+        }
+    }
```
I would like less code to live in the root of `rustc_abi`, so please move these implementations back into another file, even if this means the implementations for one type are split across multiple files. Rust allows that.

workingjubilee commented on compiler/rustc_abi/src/layout.rs:
```diff
@@ -12,12 +13,16 @@ use crate::{
 #[cfg(feature = "nightly")]
 pub use ty::{FIRST_VARIANT, FieldIdx, Layout, TyAbiInterface, TyAndLayout, VariantIdx};
 
+const MAX_SIMD_LANES: u64 = crate::TargetDataLayout::MAX_SIMD_LANES;
```
why?

workingjubilee reviewed (changes requested):
I like things overall but I'm a bit perplexed by a few changes.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13737244798/job/38422423861) [(plain)](https://github.com/rust-lang/rust/commit/578f3b947c030d460d4b45e2384c936749275c6d/checks/38422423861/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting stable_mir v0.1.0-preview (/checkout/compiler/stable_mir)
error: unresolved link to `layout_of_struct_or_enum`
   --> compiler/rustc_abi/src/layout.rs:207:47
    |
207 |     /// This uses dedicated code instead of [`layout_of_struct_or_enum`], as coroutine fields
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

 Documenting rustc_span v0.0.0 (/checkout/compiler/rustc_span)
error: could not document `rustc_abi`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Sat Mar  8 12:18:19 UTC 2025
  network time: Sat, 08 Mar 2025 12:18:19 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

workingjubilee reviewed (approved):
Thanks!

workingjubilee commented:
@bors r+

bors commented:
:pushpin: Commit 08530d3e99dd84eb0a32a00168a11cab44d70118 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 08530d3e99dd84eb0a32a00168a11cab44d70118 -->
<!-- homu: {"type":"Approved","sha":"08530d3e99dd84eb0a32a00168a11cab44d70118","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
@rust-timer build ffbe058ce34fba62f878ef42cad34656bdf03711

rust-timer commented:
Queued ffbe058ce34fba62f878ef42cad34656bdf03711 with parent 4f521991945886709a875ba2aeaa859574126c0e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4f521991945886709a875ba2aeaa859574126c0e&end=ffbe058ce34fba62f878ef42cad34656bdf03711).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~3.6 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (ffbe058ce34fba62f878ef42cad34656bdf03711): [comparison URL](https://perf.rust-lang.org/compare.html?start=4f521991945886709a875ba2aeaa859574126c0e&end=ffbe058ce34fba62f878ef42cad34656bdf03711&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4f521991945886709a875ba2aeaa859574126c0e&end=ffbe058ce34fba62f878ef42cad34656bdf03711&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4f521991945886709a875ba2aeaa859574126c0e&end=ffbe058ce34fba62f878ef42cad34656bdf03711&stat=max-rss)
<details>
<summary>Results (primary 4.4%, secondary 3.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.4% | [4.3%, 4.5%] | 2     |
| Regressions ❌ <br /> (secondary)  | 3.2% | [0.9%, 5.1%] | 28    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.4% | [4.3%, 4.5%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4f521991945886709a875ba2aeaa859574126c0e&end=ffbe058ce34fba62f878ef42cad34656bdf03711&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=4f521991945886709a875ba2aeaa859574126c0e&end=ffbe058ce34fba62f878ef42cad34656bdf03711&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 765.417s -> 765.492s (0.01%)
**Artifact size**: 362.05 MiB -> 362.03 MiB (-0.00%)
